File: prototype-rust/progred_graph/src/gid.rs
use crate::id::Id;
use im::HashMap;

pub trait Gid {
    fn edges(&self, entity: &Id) -> Option<&HashMap<Id, Id>>;

    fn get(&self, entity: &Id, label: &Id) -> Option<&Id> {
        self.edges(entity)?.get(label)
    }
}

impl<T: Gid + ?Sized> Gid for &T {
    fn edges(&self, entity: &Id) -> Option<&HashMap<Id, Id>> {
        (**self).edges(entity)
    }
}

pub struct StackedGid<Top: Gid, Bottom: Gid> {
    top: Top,
    bottom: Bottom,
}

impl<Top: Gid, Bottom: Gid> StackedGid<Top, Bottom> {
    pub fn new(top: Top, bottom: Bottom) -> Self {
        Self { top, bottom }
    }
}

impl<Top: Gid, Bottom: Gid> Gid for StackedGid<Top, Bottom> {
    fn edges(&self, entity: &Id) -> Option<&HashMap<Id, Id>> {
        self.top.edges(entity).or_else(|| self.bottom.edges(entity))
    }
}

File: progred/progred/Graph/Gid.swift
import Foundation
import HashTreeCollections

struct Edges {
    let data: TreeDictionary<Id, Id>
    let readOnly: Bool

    subscript(_ label: Id) -> Id? { data[label] }
}

protocol Gid {
    func edges(entity: Id) -> Edges?
    func get(entity: Id, label: Id) -> Id?
}

extension Gid {
    func get(entity: Id, label: Id) -> Id? {
        edges(entity: entity)?[label]
    }
}

struct PrimitiveGid: Gid {
    let recordField: Id
    let stringRecord: Id
    let numberRecord: Id

    func edges(entity: Id) -> Edges? {
        switch entity {
        case .string: Edges(data: [recordField: stringRecord], readOnly: true)
        case .number: Edges(data: [recordField: numberRecord], readOnly: true)
        case .uuid: nil
        }
    }
}

struct StackedGid<Top: Gid, Bottom: Gid>: Gid {
    let top: Top
    let bottom: Bottom

    func edges(entity: Id) -> Edges? {
        top.edges(entity: entity) ?? bottom.edges(entity: entity)
    }
}
